{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "150fc267-2f7d-4b5a-a393-43d20c4dd42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the dataset\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from sklearn.metrics import normalized_mutual_info_score\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import pickle\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy.ndimage import gaussian_filter1d\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "from tqdm import tqdm\n",
    "from matplotlib import cm\n",
    "from scipy.cluster.hierarchy import linkage, dendrogram,fcluster\n",
    "import networkx as nx\n",
    "from scipy.spatial import distance\n",
    "from scipy.spatial.distance import squareform\n",
    "from scipy.spatial.distance import pdist\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy.ndimage import gaussian_filter\n",
    "from skimage import measure\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "import zarr\n",
    "# import tarfile\n",
    "import io\n",
    "import anndata as ad\n",
    "import os"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a9d09b59-83f5-4288-81a6-3ba7561791b4",
   "metadata": {},
   "source": [
    "From Langlieb et al. 2023 (Macoscko lab)\n",
    "\n",
    "Single_Nuc_Cluster_Avg_Expression.csv.gz\n",
    "For each cluster and each gene, gives the average counts for the member nuclei\n",
    "Note that this is a large file because of the float decimal points. It is\n",
    "more accurate and faster to use the Sum_Counts above and divide each cell\n",
    "type by the `num_cells_postQC` column of the CelType_Metadata, but this is\n",
    "given for convienence.\n",
    "\n",
    "Changes to RCTD for decomposition assignment\n",
    "To aid in mapping large references with many similar clusters, we modified how RCTD scores explicit pairs of cell types in doublet mode. Rather than using the result of the single-cell type pair that fit best, we identified the cell type pairs that scored similar to the best-scoring pair (with likelihood score within 30). Then, we collated the frequency of each cell type occurring in these well-fitting pairs and divided by the total occurrences of all the cell types to make a confidence score. Throughout the paper, we use 0.3 (of a maximum score of 0.5) as the threshold for a ‘confident’ mapping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d52a1c47-74a1-42a0-8b2e-4e94e5faf6dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "df53a1cc-0e4b-41c0-99cf-74990d233909",
   "metadata": {},
   "source": [
    "## Position cell type transcriptomic centroids onto STaligned Slide-seq data from Langlieb et al. with their prefiltering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30cca75-13ed-4a9e-994a-be1417d676be",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for file_select in tqdm(os.listdir('./zenodo/csv/macosko_STalign/')):\n",
    "    puck_num_select = int(file_select.split('_')[-1].split('.')[0])\n",
    "    section_select = int(file_select.split('__')[0].split('_')[1])\n",
    "    \n",
    "    print(section_select)\n",
    "\n",
    "    # load the scRNA-seq centroid transcriptomes\n",
    "    dfCluster_average_Expression = pd.read_csv('/data/LBA_DATA/Macosko/Single_Nuc_Cluster_Avg_Expression.csv.gz', index_col=0)\n",
    "    dfCluster_average_Expression.columns = [_.split('=')[0] for _ in dfCluster_average_Expression.columns]\n",
    "\n",
    "    # prepare the puck with STAligned coordinates\n",
    "    puck_name = f'Puck_Num_{puck_num_select}'\n",
    "    puck = ad.read_h5ad(f\"/data/LBA_DATA/Macosko/All_Pucks_h5ad/{puck_name}.h5ad\")\n",
    "    matrix_file = f'/data/LBA_DATA/Macosko/Mapping_Matrices/{puck_name}.mapping.mtx.gz'\n",
    "    matrix = scipy.io.mmread(matrix_file)\n",
    "    metadata_file = f'/data/LBA_DATA/Macosko/Mapping_Matrices/{puck_name}.mapping.metadata.tsv.gz'\n",
    "    metadata = pd.read_csv(metadata_file, sep='\\t', index_col=0)\n",
    "    cell_types_file = f'/data/LBA_DATA/Macosko/Mapping_Matrices/{puck_name}.mapping.MappedCellTypes.txt'\n",
    "    cell_types = np.loadtxt(cell_types_file, dtype=str)\n",
    "    ctass= pd.DataFrame(matrix.toarray(), columns=cell_types)\n",
    "    ctass.index = metadata.index\n",
    "    ctass['puckID'] = puck_name\n",
    "    c1 = pd.read_csv(f\"/data/LBA_DATA/Macosko/macosko_STalign/MSI_{section_select}__Macosko_{puck_num_select}.csv\", index_col=0)\n",
    "\n",
    "    # prefilter the cell types puck using spatial matching to our correspondent section\n",
    "    puck_triplets = puck.obs[['CCF_Y', 'CCF_Z', 'CCF_X']]\n",
    "    c1_triplets = c1[['y_index', 'z_index', 'x_index']]\n",
    "    puck_tuples = list(puck_triplets.itertuples(index=False, name=None))\n",
    "    c1_tuples = set(c1_triplets.itertuples(index=False, name=None))\n",
    "    mask = [triplet in c1_tuples for triplet in puck_tuples]\n",
    "    puck_filtered = puck[mask].copy()\n",
    "    c1['acronym'] = np.array(puck_filtered.obs['CCF_acronym'])\n",
    "    c1['puckID'] = np.array(puck_filtered.obs.index)\n",
    "    c1 = c1.set_index('puckID').loc[ctass.index]\n",
    "\n",
    "    # drop the columns that are not in the snRNA data\n",
    "    celltypes_intersect = np.intersect1d(ctass.columns[:-1], dfCluster_average_Expression.index)\n",
    "    ctass = ctass.loc[:,celltypes_intersect]\n",
    "    dfCluster_average_Expression = dfCluster_average_Expression.loc[ctass.columns,:]\n",
    "\n",
    "    # follow Langlieb et al guidelines on RCTD believable cell type assignments\n",
    "    ctass[ctass < 0.3] = 0.0\n",
    "    ctass = ctass>0\n",
    "    ctass = ctass.loc[ctass.sum(axis=1) == 1,:]\n",
    "\n",
    "    # assign the scRNA-seq centroid for each cell type singlet found by Slide-seq\n",
    "    active_clusters = ctass.astype(int).idxmax(axis=1)\n",
    "    cell_x_gene = dfCluster_average_Expression.loc[active_clusters].copy()\n",
    "    cell_x_gene.index = ctass.index\n",
    "    cell_x_gene['y_index'] = c1.loc[cell_x_gene.index, 'y_index_new']\n",
    "    cell_x_gene['z_index'] = c1.loc[cell_x_gene.index, 'z_index_new']\n",
    "\n",
    "    # save the imputed gene expression puck to file\n",
    "    cell_x_gene_avg = cell_x_gene.groupby(level=0, axis=1).mean()\n",
    "    output_filename = f\"gexpr_{section_select}.h5\"\n",
    "    cell_x_gene_avg.to_hdf(output_filename, key=\"table\", mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4998d96-c96c-4ed2-b5b3-acac21b0addf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e18c3807-763b-4d30-8805-bcabfebce75d",
   "metadata": {},
   "source": [
    "## Prepare to compare cell types and their neighbor lipizones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f9b517a-5f67-46c4-a4f7-ddd118d6e509",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_results = []\n",
    "\n",
    "for i in tqdm(range(1, 33)):\n",
    "    try:\n",
    "        gexpr = pd.read_hdf(f\"./gene_expression/gexpr_{i}.h5\", key=\"table\")\n",
    "        \n",
    "        sec = dat.loc[dat['Section'] == i,:]\n",
    "        sec = sec.loc[sec['z_index'] > 456/2,:]\n",
    "        \n",
    "        distance_matrix = cdist(gexpr[['y_index', 'z_index']], \n",
    "                              sec[['y_index', 'z_index']], \n",
    "                              metric='euclidean')\n",
    "        distance_df = pd.DataFrame(distance_matrix, \n",
    "                                 index=gexpr.index, \n",
    "                                 columns=sec.index)\n",
    "        distance_df_masked = distance_df.copy()\n",
    "        distance_df_masked[distance_df_masked > 4] = np.nan # this is still relatively large as a radius\n",
    "        \n",
    "        # find closest cells and their indices\n",
    "        closest_indices = distance_df_masked.idxmin(axis=0)\n",
    "\n",
    "        all_results.append(closest_indices)\n",
    "        del gexpr, distance_df_masked, distance_df\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Error processing section {i}: {str(e)}\")\n",
    "        continue\n",
    "\n",
    "pixelclosestcells = pd.concat(all_results)\n",
    "pixelclosestcells.to_hdf(\"pixelclosestcells.h5ad\", key=\"table\")\n",
    "pixelclosestcells"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
